runtime/domain.c

```diff
@@ -2101,15 +2145,97 @@ static void domain_terminate (void)
   /* signal the domain termination to the backup thread
      NB: for a program with no additional domains, the backup thread
      will not have been started */
-  atomic_store_release(&domain_self->backup_thread_msg, BT_TERMINATE);
-  caml_plat_signal(&domain_self->domain_cond);
+  terminate_backup_thread(domain_self);
   caml_plat_unlock(&domain_self->domain_lock);
 
   caml_plat_assert_all_locks_unlocked();
   /* This is the last thing we do because we need to be able to rely
      on caml_domain_alone (which uses caml_num_domains_running) in at least
      the shared_heap lockfree fast paths */
-  atomic_fetch_add(&caml_num_domains_running, -1);
+  if (!last)
```

Comment:
Before this PR, `domain_terminate` is only invoked by domain threads, but never by the first domain. With this PR, the first domain invokes `domain_terminate` to perform more cleaning, but if it decrements the counter, then it would have a value of zero and `caml_domain_alone()` would return `false`. I guess I could change `caml_domain_alone` to use `<= 1` rather than `== 1`...

Reply:
Either way, it needs a comment. I am leaning towards leaving the code as is and adding a comment rather than changing the implementation of `caml_domain_alone` where (a) we'll have to change the nice equality check with the inequality and (b) add a comment explaining why `caml_num_domains_running` could be `0`.

Reply:
Done. I'll squash these last few commits into the earlier commits they belong to, once you're happy with the changes.